### ISSUE TITLE
associativity proof depends on scalar coefficients?

### ISSUE THREAD

**Issue by Infinoid** [halide/Halide#4853]
I am having some problems with rfactor, and I managed to reduce it down to this simple program:

```
#include <Halide.h>
using namespace Halide;
#include "../tools/halide_image_io.h"
using namespace Halide::Tools;

int main(int argc, char **argv) {

    Halide::Buffer<uint8_t> in = load_image("rgb.png");
    Var i{"i"}, j{"j"}, rx{"rx"}, ry{"ry"}, rz{"rz"};
    RDom r{0, 8, 0, 8, 0, 3, "r"};
    Func f{"f"};
    f(i,j) = Expr((double)0.0);
    f(r.x,r.y) += in(r.x, r.y, r.z) * Expr(-5.0);

    Func f_intm = f.update(0).rfactor({{r.x, rx}, {r.y, ry}});

    Buffer<double> output = f.realize(100, 100);
}
```

When you compile and run it, it dies with the following error:

```
Error:
Failed to call rfactor() on f.update(0) since it can't prove associativity of the operator
Aborted
```

If you remove the negative sign, changing `-5.0` to `5.0`, it completes successfully.

Is this a bug?  I don't understand why rfactor would care what the scalar coefficient is.

**Comment by abadams** [comment]
Yes this is a bug. Looks like it gets simplified to f(r.x, r.y) -= in(r.x, r.y, r.z) * 5.0, and then because subtraction isn't an associative/commutative binary operator it gives up. Thinking about how to fix...

**Comment by abadams** [comment]
Actually that "simplification" is in the process of being removed (hopefully) in #4857 , so += should not get rewritten to -= in future. 

Would be better if we could recognize -= as a valid summation too, just with the output sign flipped, but rfactor can't currently do that sort of rewrite.

**Comment by Infinoid** [comment]
After #4857 was merged, the original test program runs correctly now.  But as @abadams mentioned, associativity still fails for -=.  These two lines should be equivalent:

```
    f(r.x,r.y) += in(r.x, r.y, r.z) * Expr(-5.0);
    f(r.x,r.y) -= in(r.x, r.y, r.z) * Expr(5.0);
```

But the second one fails to compile.  `Error: Failed to call rfactor() on f.update(0) since it can't prove associativity of the operator`
